integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/automaticindexing/AutomaticIndexingSortedMapValuesAssociationIT.java

```diff
@@ -0,0 +1,519 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.integrationtest.mapper.orm.automaticindexing;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import java.util.stream.Stream;
+import javax.persistence.Basic;
+import javax.persistence.ElementCollection;
+import javax.persistence.Entity;
+import javax.persistence.Id;
+import javax.persistence.JoinColumn;
+import javax.persistence.JoinTable;
+import javax.persistence.ManyToMany;
+import javax.persistence.MapKeyColumn;
+import javax.persistence.OneToOne;
+import javax.persistence.OrderBy;
+import javax.persistence.Transient;
+
+import org.hibernate.annotations.SortNatural;
+import org.hibernate.search.mapper.pojo.dirtiness.ReindexOnUpdate;
+import org.hibernate.search.mapper.pojo.mapping.definition.annotation.Field;
+import org.hibernate.search.mapper.pojo.mapping.definition.annotation.Indexed;
+import org.hibernate.search.mapper.pojo.mapping.definition.annotation.IndexedEmbedded;
+import org.hibernate.search.mapper.pojo.mapping.definition.annotation.IndexingDependency;
+import org.hibernate.search.mapper.pojo.mapping.definition.annotation.ObjectPath;
+import org.hibernate.search.mapper.pojo.mapping.definition.annotation.PropertyValue;
+import org.hibernate.search.util.impl.test.annotation.TestForIssue;
+
+/**
+ * Test automatic indexing based on Hibernate ORM entity events
+ * when a SortedMap-values association is involved.
+ * <p>
+ * See {@link AbstractAutomaticIndexingAssociationIT} for more details on how this test is designed.
+ */
+@TestForIssue(jiraKey = "HSEARCH-2490")
+public class AutomaticIndexingSortedMapValuesAssociationIT extends AbstractAutomaticIndexingMultiAssociationIT<
+		AutomaticIndexingSortedMapValuesAssociationIT.IndexedEntity,
+		AutomaticIndexingSortedMapValuesAssociationIT.ContainingEntity,
+		AutomaticIndexingSortedMapValuesAssociationIT.ContainedEntity,
+		SortedMap<String, AutomaticIndexingSortedMapValuesAssociationIT.ContainedEntity>,
+		List<AutomaticIndexingSortedMapValuesAssociationIT.ContainingEntity>
+		> {
+
+	public AutomaticIndexingSortedMapValuesAssociationIT() {
+		super( new SortedMapValuesAssociationModelPrimitives() );
+	}
+
+	private static class SortedMapValuesAssociationModelPrimitives
+			implements MultiAssociationModelPrimitives<IndexedEntity, ContainingEntity, ContainedEntity,
+					SortedMap<String, ContainedEntity>, List<ContainingEntity>> {
```

Comment:
We already do, actually. This class contains the bare minimum (the "primitives"): all the methods that must refer directly to the domain model. All the test methods are in the abstract test class.

I can't use abstract classes in the domain model because that would introduce a bias in our tests, during bootstrap in particular: Hibernate Search doing reflection over a class that extends a generic one is not the same as Hibernate Search  doing reflection over a top-level (non-extending) class. And there would probably be other changes, such as how Hibernate ORM builds its internal model of such classes, how properties are accessed (when the abstract class defines the getters, you probably won't be able to use direct field access), and so on. So yeah... I'd rather not go any further regarding test code factorizing.